sdk/src/main/scala/com/horizen/validation/ConsensusValidator.scala

```diff
@@ -71,42 +71,69 @@ class ConsensusValidator extends HistoryBlockValidator with ScorexLogging {
-                                     previousFullConsensusEpochInfo: FullConsensusEpochInfo,
-                                     history: SidechainHistory): Unit = {
+                                    currentFullConsensusEpochInfo: FullConsensusEpochInfo,
+                                    previousFullConsensusEpochInfo: FullConsensusEpochInfo,
+                                    bestKnownParentId: ModifierId,
+                                    bestKnownParentInfo: SidechainBlockInfo,
+                                    history: SidechainHistory,
+                                    previousEpochOmmersInfoAccumulator: Seq[(VrfOutput, ConsensusSlotNumber)]
+                                   ): Unit = {
     val ommers = ommersContainer.ommers
-    if (ommers.isEmpty)
+    if(ommers.isEmpty)
       return
 
     val ommersContainerEpochNumber: ConsensusEpochNumber = history.timeStampToEpochNumber(ommersContainer.header.timestamp)
 
-    // Ommers can be from the same ConsensusEpoch as OmmersContainer or from previous epoch.
-    // If first part of Ommers is from previous epoch, they can have data, that will lead to different FullConsensusEpochInfo for current epoch.
-    // So the second part of Ommers should be verified against different FullConsensusEpochInfo: nonce part can be different
-    // With current Nonce calculation algorithm, it's not possible to retrieve MainchainHeaders with RefData, so no way to recalculate proper Nonce.
-    // Solutions: Nonce must be taken not from the whole MCRefs most PoW header, but like in original Ouroboros from the VRF, that is a part of SidechainBlockHeader.
-    var isPreviousEpochOmmer: Boolean = false
+    var accumulator: Seq[(VrfOutput, ConsensusSlotNumber)] = previousEpochOmmersInfoAccumulator
+    var previousOmmerEpochNumber: ConsensusEpochNumber = ommersContainerEpochNumber
+    var ommerFullConsensusEpochInfo = currentFullConsensusEpochInfo
+
     for(ommer <- ommers) {
-      val ommerEpochNumber: ConsensusEpochNumber = history.timeStampToEpochNumber(ommer.header.timestamp)
-      // Fork occurs in previous consensus epoch
-      if(ommerEpochNumber < ommersContainerEpochNumber) {
-        isPreviousEpochOmmer = true
-        val vrfOutput = history.getVrfOutput(ommer.header, previousFullConsensusEpochInfo.nonceConsensusEpochInfo)
-          .getOrElse(throw new IllegalStateException(s"VRF check for block ${ommer.header.id} had been failed"))
-        verifyForgerBox(ommer.header, previousFullConsensusEpochInfo.stakeConsensusEpochInfo, vrfOutput)
-        verifyOmmers(ommer, previousFullConsensusEpochInfo, null, history)
+      val ommerEpochAndSlot: ConsensusEpochAndSlot = history.timestampToEpochAndSlot(ommer.header.timestamp)
+
+      if(ommerEpochAndSlot.epochNumber < previousOmmerEpochNumber) {
+        // First ommer is from previous consensus epoch to Ommer Container epoch.
+        ommerFullConsensusEpochInfo = previousFullConsensusEpochInfo
```

Comment:
so when calling later `verifyOmmers(ommer, ommerFullConsensusEpochInfo, previousFullConsensusEpochInfo,	        bestKnownParentId, bestKnownParentInfo, history, accumulator)` we will pass the same data for `ommerFullConsensusEpochInfo` and for `ommerFullConsensusEpochInfo`. It is not an issue, but in such case we have inconsistency. I think it is a good style to use Option for `previousFullConsensusEpochInfo` then to avoid it.

Reply:
Agree.
Let's expect `previousFullConsensusEpochInfo: Option[...]` and add `ommerPreviousFullConsensusEpochInfo: Option[...]` after line 108, so will set/modify it properly and pass it for subommers validation.